go_lib/hooks/generate_password/hook.go

```diff
@@ -17,53 +17,211 @@ limitations under the License.
+				},
+				NamespaceSelector: &types.NamespaceSelector{
+					NameSelector: &types.NameSelector{
+						MatchNames: []string{hook.Secret.Namespace},
+					},
+				},
+				// Synchronization is redundant because of OnBeforeHelm.
+				ExecuteHookOnSynchronization: go_hook.Bool(false),
+				ExecuteHookOnEvents:          go_hook.Bool(false),
+				FilterFunc:                   hook.Filter,
+			},
+		},
+		OnBeforeHelm: &go_hook.OrderedConfig{Order: float64(hook.BeforeHelmOrder)},
+	}, hook.Handle)
 }
 
-func generatePasswordWithArgs(input *go_hook.HookInput, moduleValuesPath string) error {
-	if input.Values.Exists(fmt.Sprintf("%s.auth.externalAuthentication", moduleValuesPath)) {
-		input.ConfigValues.Remove(fmt.Sprintf("%s.auth.password", moduleValuesPath))
-		if input.ConfigValues.Exists(fmt.Sprintf("%s.auth", moduleValuesPath)) && len(input.ConfigValues.Get(fmt.Sprintf("%s.auth", moduleValuesPath)).Map()) == 0 {
-			input.ConfigValues.Remove(fmt.Sprintf("%s.auth", moduleValuesPath))
-		}
+type Hook struct {
+	Secret          Secret
+	Values          Values
+	BeforeHelmOrder int
 
-		return nil
+	PasswordGeneratorFunc func(input *go_hook.HookInput) string
+}
+
+type Secret struct {
+	Namespace string
+	Name      string
+
+	RawPasswordField    string
+	BasicAuthPlainField string
+
+	FilterFunc func(secret *v1.Secret) (go_hook.FilterResult, error)
+}
+
+type Values struct {
+	ModuleKey           string
+	PasswordKey         string
+	PasswordInternalKey string
+	ExternalAuthKey     string
+}
+
+// Filter extracts password from the Secret. Password can be stored as a raw string or as
+// a basic auth plain format (user:{PLAIN}password). Custom FilterFunc is called for custom
+// password extraction.
+func (h *Hook) Filter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	secret := &v1.Secret{}
+	err := sdk.FromUnstructured(obj, secret)
+	if err != nil {
+		return nil, fmt.Errorf("cannot convert secret to struct: %v", err)
 	}
 
-	if input.Values.Exists(fmt.Sprintf("%s.auth.password", moduleValuesPath)) {
-		return nil
+	if h.Secret.FilterFunc != nil {
+		return h.Secret.FilterFunc(secret)
 	}
 
-	if !input.ConfigValues.Exists(fmt.Sprintf("%s.auth", moduleValuesPath)) {
-		input.ConfigValues.Set(fmt.Sprintf("%s.auth", moduleValuesPath), json.RawMessage("{}"))
+	if h.Secret.RawPasswordField != "" {
+		return string(secret.Data[h.Secret.RawPasswordField]), nil
 	}
 
-	generatedPass := pwgen.AlphaNum(20)
+	// Return field with basic auth.
+	return string(secret.Data[h.SecretField()]), nil
+}
 
-	input.ConfigValues.Set(fmt.Sprintf("%s.auth.password", moduleValuesPath), generatedPass)
+// Handle restores password from the configuration or from the Secret and
+// puts it to internal values.
+// It generates new password if no password found in the configuration and the
+// Secret or no externalAuthentication defined.
+func (h *Hook) Handle(input *go_hook.HookInput) error {
+	var externalAuthKey = h.ExternalAuthKey()
```

Comment:
```suggestion
	externalAuthKey := h.ExternalAuthKey()
```
It is better to stick to the single way of variables declaration.

Reply:
Done